unified-runtime/source/common/logger/ur_sinks.hpp

```
@@ -211,6 +214,36 @@ class FileSink : public Sink {
   std::ofstream ofstream;
 };
 
+class CallbackSink : public Sink {
+public:
+  CallbackSink(std::string logger_name, bool skip_prefix = false,
+               bool skip_linebreak = false)
+      : Sink(std::move(logger_name), skip_prefix, skip_linebreak) {}
+
+  CallbackSink(std::string logger_name, ur_logger_level_t flush_lvl,
+               bool skip_prefix, bool skip_linebreak)
+      : CallbackSink(std::move(logger_name), skip_prefix, skip_linebreak) {
+    this->flush_level = flush_lvl;
+  }
+
+  ~CallbackSink() = default;
+
+  void setCallback(ur_logger_callback_t cb, void *pUserData) {
+    callback = cb;
+    userData = pUserData;
+  }
+
+private:
+  ur_logger_callback_t callback = nullptr;
+  void *userData = nullptr;
+
+  virtual void print(ur_logger_level_t level, const std::string &msg) override {
+    if (callback) {
```

should we maybe do some kind of die/assert if `callback` has been set to an invalid value?

---

how would you detect its invalid?

---

just at minimum checking if its null and informing the user in some way? currently it would fail silently since it will only get called if its not null

---

We should be checking if the callback is null when initializing the callback. After that point, we should be able assume it's not null.

I'd hope the sink is never invoked unless it's been enabled. Is that how it works?